zero_noise_extrapolation_cnot.py

```diff
@@ -0,0 +1,570 @@
+
+        exp_vals = self.compute_exp_vals(counts)
+
+        # Process the resulting expectation values:
+
+        self.bare_exp_vals = zeros((repeats,))
+        self.all_exp_vals = zeros((repeats,n_amp_factors))
+        self.mitigated_exp_vals = zeros((repeats,))
+
+        if verbose:
+            print("Processing results")
+
+        if repeats == 1:
+            self.bare_exp_vals[0] = exp_vals[0]
+            self.all_exp_vals[0,:] = exp_vals
+
+            self.result = richardson_extrapolate(self.all_exp_vals[0,:], self.noise_amplification_factors)
+
+            self.mitigated_exp_vals[0] = self.result
+        else:
+            for i in range(repeats):
+                self.bare_exp_vals[i] = exp_vals[i*n_amp_factors]
+
+                for j in range(n_amp_factors):
+                    self.all_exp_vals[i, j] = exp_vals[i*n_amp_factors + j]
+
+                mitigation_results = richardson_extrapolate(self.all_exp_vals[i, :], self.noise_amplification_factors)
+                self.mitigated_exp_vals[i] = mitigation_results
+
+            self.result = sum(self.mitigated_exp_vals) / shape(self.mitigated_exp_vals)[0]
+
+        if verbose:
+            print("Mitigation done. Result:")
+            print("Bare exp val =", average(self.bare_exp_vals))
+            print("Mitigated exp val =",self.result)
+
+        return self.result
+
+
+# PAULI TWIRLING AND NOISE AMPLIFICATION HELP FUNCTIONS
+
+# Conversion from pauli x/y/z-gates to physical u1/u3-gates in correct OpenQASM-format
+PHYSICAL_GATE_CONVERSION = {"X": "u3(pi,0,pi)", "Z": "u1(pi)", "Y": "u3(pi,pi/2,pi/2)"}
+
+
+def find_qreg_name(circuit_qasm: str) -> str:
+    """
+    Finds the name of the quantum register in the circuit.
+
+    :param circuit_qasm: OpenQASM string with instructions for the entire circuit
+    :return: Name of the quantum register
+    """
+    for line in circuit_qasm.splitlines():
+        if line[0:5] == "qreg ":
+            qreg_name = ""
+            for i in range(5,len(line)):
+                if line[i] == "[" or line[i] == ";":
+                    break
+                elif line[i] != " ":
+                    qreg_name += line[i]
+            return qreg_name
+
+
+def find_cnot_control_and_target(qasm_line: str) -> (int, int):
+    """
+    Find indices of control and target qubits for the CNOT-gate in question
+
+    :param qasm_line: OpenQASM line containing the CNOT
+    :return: Indices of control and target qubits
+    """
+    qubits = []
+    for i, c in enumerate(qasm_line):
+        if c == "[":
+            qubit_nr = ""
+            for j in range(i+1, len(qasm_line)):
+                if qasm_line[j] == "]":
+                    break
+                qubit_nr += qasm_line[j]
+            qubits.append(int(qubit_nr))
+    return qubits[0], qubits[1]
+
+
+def propagate(control_in: str, target_in: str):
+    """
+    Propagates Pauli gates through a CNOT in accordance with the following circuit identities:
```

Comment:
Are you trying to remove Pauli gates? If yes, I think this will not result in a twirl.